pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,494 @@
+		leaseClone.Spec.PreferredHolder = &electeeName
+		_, err = c.leaseClient.Leases(leaderLeaseID.namespace).Update(ctx, leaseClone, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func pickBestStrategy(candidates []*v1alpha1.LeaseCandidate) v1.CoordinatedLeaseStrategy {
+	// TODO: This doesn't account for cycles within the preference graph
+	// We may have to do a topological sort to verify that the preference ordering is valid
+	var bestStrategy *v1.CoordinatedLeaseStrategy
+	for _, c := range candidates {
+		if len(c.Spec.PreferredStrategies) > 0 {
+			if bestStrategy == nil {
+				bestStrategy = &c.Spec.PreferredStrategies[0]
+				continue
+			}
+			if *bestStrategy != c.Spec.PreferredStrategies[0] {
+				if idx := slices.Index(c.Spec.PreferredStrategies, *bestStrategy); idx > 0 {
+					bestStrategy = &c.Spec.PreferredStrategies[0]
+				} else {
+					klog.Infof("Error: bad strategy ordering")
+				}
+			}
+		}
+	}
+	return (*bestStrategy)
+}
+
+func pickBestLeader(candidates []*v1alpha1.LeaseCandidate) *v1alpha1.LeaseCandidate {
+	var electee *v1alpha1.LeaseCandidate
+	for _, c := range candidates {
+		if electee == nil || compare(electee, c) > 0 {
+			electee = c
+		}
+	}
+	if electee == nil {
+		klog.Infof("pickBestLeader: none found")
+	} else {
+		klog.Infof("pickBestLeader: %s %s", electee.Namespace, electee.Name)
+	}
+	return electee
+}
+
+func shouldReelect(candidates []*v1alpha1.LeaseCandidate, currentLeader *v1alpha1.LeaseCandidate) bool {
+	klog.Infof("shouldReelect for candidates: %+v", candidates)
+	pickedLeader := pickBestLeader(candidates)
+	if pickedLeader == nil {
+		return false
+	}
+	return compare(currentLeader, pickedLeader) > 0
+}
+
+func getEmulationVersion(l *v1alpha1.LeaseCandidate) semver.Version {
+	value := l.Spec.EmulationVersion
+	v, err := semver.ParseTolerant(value)
+	if err != nil {
+		return semver.Version{}
+	}
+	return v
+}
+
+func getBinaryVersion(l *v1alpha1.LeaseCandidate) semver.Version {
+	value := l.Spec.BinaryVersion
+	v, err := semver.ParseTolerant(value)
+	if err != nil {
+		return semver.Version{}
+	}
+	return v
+}
+
+// -1: lhs better, 1: rhs better
+func compare(lhs, rhs *v1alpha1.LeaseCandidate) int {
+	lhsVersion := getEmulationVersion(lhs)
+	rhsVersion := getEmulationVersion(rhs)
+	result := lhsVersion.Compare(rhsVersion)
+	if result == 0 {
+		lhsVersion := getBinaryVersion(lhs)
+		rhsVersion := getBinaryVersion(rhs)
+		result = lhsVersion.Compare(rhsVersion)
+	}
+	if result == 0 {
+		if lhs.CreationTimestamp.After(rhs.CreationTimestamp.Time) {
+			return 1
+		}
+		return -1
+	}
+	return result
+}
+
+func (c *Controller) listAdmissableCandidates(leaderLeaseID leaderLeaseID) ([]*v1alpha1.LeaseCandidate, error) {
+	leases, err := c.leaseCandidateInformer.Lister().LeaseCandidates(leaderLeaseID.namespace).List(labels.Everything())
```

Comment:
TODO, you can optimize this by using a label on the LeaseCandidates so it return directly the list of the candidates

Reply:
Does it make sense to create a label when there is a field already present? We can't perform validation on the label so a user changing it could cause the controller to be in a weird state.